File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
    vector<vector<int>> ans;
public:
    void backtracking(vector<int>& candidates, int target, vector<int> &temps ,int i = 0) {
        
        if(target == 0){
            ans.push_back(temps);
            return;
        }
        while(i < candidates.size() && target - candidates[i] >= 0) {

            temps.push_back(candidates[i]);
            
            backtracking(candidates, target - candidates[i], temps, i);
            
            i++;

            temps.pop_back();
        }

    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        sort(candidates.begin(), candidates.end());

//         auto last = std::unique(candidates.begin(), candidates.end());
//         candidates.erase(last, candidates.end());
        backtracking(candidates, target, temp, 0);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        n = len(candidates)
        ans = []
        candidates.sort()
        
        def solve(total, curr, res):
            if total < 0: return
            if total == 0:
                ans.append(res[:])
                return
            
            for i in range(curr, n):
                if (target - candidates[i] < 0 or candidates[i] > target):
                    continue
                res.append(candidates[i])
                solve(total - candidates[i], i, res)
                res.pop()
        solve(target, 0, [])
        return ans
            